Documentation/Proposals/NNNN-custom-test-execution-traits.md

```diff
@@ -0,0 +1,376 @@
+to resolve that is to change it to a `@TaskLocal` variable, as this would be
+concurrency-safe and still allow tests accessing this state to run in parallel:
+
+```swift
+extension APICredentials {
+  @TaskLocal static var current: Self?
+}
+```
+
+Binding task local values requires using the scoped access
+[`TaskLocal.withValue()`](https://developer.apple.com/documentation/swift/tasklocal/withvalue(_:operation:isolation:file:line:)
+API though, and that would not be possible if `Trait` exposed separate methods
+like `setUp()` and `tearDown()`.
+
+For these reasons, I believe it's important to expose this trait capability
+using a single, scoped access-style API which accepts a closure. A simplified
+version of that idea might look like this:
+
+```swift
+public protocol Trait: Sendable {
+  // ...
+
+  // Simplified example, not the actual proposal
+  func executeTest(_ body: @Sendable () async throws -> Void) async throws
+}
+
+extension MockAPICredentialsTrait {
+  func executeTest(_ body: @Sendable () async throws -> Void) async throws {
+    let mockCredentials = APICredentials(apiKey: "...")
+    try await APICredentials.$current.withValue(mockCredentials) {
+      try await body()
+    }
+  }
+}
+```
+
+### Avoiding unnecessarily lengthy backtraces
+
+A scoped access-style API has some potential downsides. To apply this approach
+to a test function, the scoped call of a trait must wrap the invocation of that
+test function, and every _other_ trait applied to that same test which offers
+custom behavior _also_ must wrap the other traits' calls in a nesting fashion.
+To visualize this, imagine a test function with multiple traits:
+
+```swift
+@Test(.traitA, .traitB, .traitC)
+func exampleTest() {
+  // ...
+}
+```
+
+If all three of those traits customize test execution behavior, then each of
+them needs to wrap the call to the next one, and the last trait needs to wrap
+the invocation of the test, illustrated by the following:
+
+```
+TraitA.executeTest {
+  TraitB.executeTest {
+    TraitC.executeTest {
+      exampleTest()
+    }
+  }
+}
+```
+
+Tests may have an arbitrary number of traits applied to them, including those
+inherited from containing suite types. A naïve implementation in which _every_
+trait is given the opportunity to customize test behavior by calling its scoped
+access API might cause unnecessarily lengthy backtraces that make debugging the
+body of tests more difficult. Or worse: if the number of traits is great enough,
+it could cause a stack overflow.
+
+In practice, most traits probably will _not_ need to customize test behavior, so
+to mitigate these downsides it's important that there be some way to distinguish
+traits which customize test behavior. That way, the testing library can limit
+these scoped access calls to only the traits which require it.
+
+## Detailed design
+
+I propose the following new APIs:
+
+- A new protocol `CustomTestExecuting` with a single required `execute(...)`
```

Comment:
Gerunds for protocol names are common in Objective-C, but in Swift we usually prefer an adjective form, e.g. (bikeshedding here!) `CustomTestExecutable` or `CustomExecutable` or `CustomTestSetuppableAndTeardownable`.

Reply:
Definitely open to these alternate suggestions, thanks. But I'll note: as I was considering names myself, I found this in the [Swift API Design Guidelines](https://www.swift.org/documentation/api-design-guidelines/#protocols-describing-capability-should-use-suffixes):

> - Protocols that describe a capability should be named using the suffixes `able`, `ible`, or `ing` (e.g. `Equatable`, `ProgressReporting`).

Based on that, I think an `ing` suffix would be acceptable.

Reply:
Oh it's not _wrong_, just uncommon. So it's worth thinking about something more idiomatic if we can come up with something.

Reply:
One thing I slightly dislike about the `able` suffixes is that it makes it seem as if the work is optional. "This type is _able_ to execute the test… but will it? Who knows!?". But if you define a trait this way, it **will** execute tests it's applied to, and to me the `ing` conveys that a little more clearly. "This type is responsible for _executing_ the test."

Reply:
We have plenty of counter-examples too where a noun is used instead: `Collection`, `Sequence`, `Actor`… maybe it really just is `CustomTestExecutor`.

Reply:
> We have plenty of counter-examples too where a noun is used instead: `Collection`, `Sequence`, `Actor`… maybe it really just is `CustomTestExecutor`.

The API Design Guidelines [address](https://www.swift.org/documentation/api-design-guidelines/#protocols-describing-what-is-should-read-as-nouns) that other common naming convention for protocols in the bullet above the one I quoted:

> * **Protocols that describe _what something is_ should read as nouns** (e.g. `Collection`).

So I think this adheres to the recommendations well: the `Trait` protocol best describes what the type _is_, and the `CustomTestExecuting` protocol describes one optional _capability_ a type (usually, a trait) can have.